src/backend/graphics/egl.rs

```diff
@@ -0,0 +1,746 @@
+                out.push(multisampling as c_int);
+            }
+
+            if reqs.stereoscopy {
+                return Err(CreationError::NoAvailablePixelFormat);
+            }
+
+            out.push(ffi::egl::NONE as c_int);
+            out
+        };
+
+        // calling `eglChooseConfig`
+        let mut config_id = mem::uninitialized();
+        let mut num_configs = mem::uninitialized();
+        if egl.ChooseConfig(display,
+                            descriptor.as_ptr(),
+                            &mut config_id,
+                            1,
+                            &mut num_configs) == 0 {
+            return Err(CreationError::OsError(String::from("eglChooseConfig failed")));
+        }
+        if num_configs == 0 {
+            return Err(CreationError::NoAvailablePixelFormat);
+        }
+
+        // analyzing each config
+        macro_rules! attrib {
+            ($egl:expr, $display:expr, $config:expr, $attr:expr) => (
+                {
+                    let mut value = mem::uninitialized();
+                    let res = $egl.GetConfigAttrib($display, $config,
+                                                   $attr as ffi::egl::types::EGLint, &mut value);
+                    if res == 0 {
+                        return Err(CreationError::OsError(String::from("eglGetConfigAttrib failed")));
+                    }
+                    value
+                }
+            )
+        };
+
+        let desc = PixelFormat {
+            hardware_accelerated: attrib!(egl, display, config_id, ffi::egl::CONFIG_CAVEAT) !=
+                                  ffi::egl::SLOW_CONFIG as i32,
+            color_bits: attrib!(egl, display, config_id, ffi::egl::RED_SIZE) as u8 +
+                        attrib!(egl, display, config_id, ffi::egl::BLUE_SIZE) as u8 +
+                        attrib!(egl, display, config_id, ffi::egl::GREEN_SIZE) as u8,
+            alpha_bits: attrib!(egl, display, config_id, ffi::egl::ALPHA_SIZE) as u8,
+            depth_bits: attrib!(egl, display, config_id, ffi::egl::DEPTH_SIZE) as u8,
+            stencil_bits: attrib!(egl, display, config_id, ffi::egl::STENCIL_SIZE) as u8,
+            stereoscopy: false,
+            double_buffer: true,
+            multisampling: match attrib!(egl, display, config_id, ffi::egl::SAMPLES) {
+                0 | 1 => None,
+                a => Some(a as u16),
+            },
+            srgb: false, // TODO: use EGL_KHR_gl_colorspace to know that
+        };
+
+        let surface = {
+            let surface = match native {
+                Native::X11(_, window) |
+                Native::Wayland(_, window) |
+                Native::Gbm(_, window) => egl.CreateWindowSurface(display, config_id, window, ptr::null()),
+            };
+
+            if surface.is_null() {
+                return Err(CreationError::OsError(String::from("eglCreateWindowSurface failed")));
+            }
+            surface
+        };
```

Comment:
Good news, I found the cause for the segfault!

The reason is, because of the reentrancy of this function(see the match on line 146, it calls itself), this par ends up called twice. Especially, `egl.CreateWindowSurface()` is called twice on the same native surface, and apparently this is not good.

The simplest fix is to move this whole `let surface = { .... }` block down in the function, right before the end, until after the EGL context has been successfully created (so line 500 of this diff). :)